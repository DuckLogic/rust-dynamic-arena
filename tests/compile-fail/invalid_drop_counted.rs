extern crate dynamic_arena;

use dynamic_arena::DynamicArena;
use std::cell::Cell;

pub struct DropCounted<'a>(&'a Cell<u32>);

fn main() {
    let arena = DynamicArena::new_bounded();
    {
        let cell = Box::new(Cell::new(0));
        for _i in 0..5 {
            arena.alloc(DropCounted(&cell));
        }
    }
    /*
     * Running this drop would reference an invalid pointer to the stack allocated `Cell`.
     * In order to maintain rust's safety guarantee this must be statically prevented.
     */
    drop(arena);
}
