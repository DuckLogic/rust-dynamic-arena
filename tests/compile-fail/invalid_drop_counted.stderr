error[E0597]: `cell` does not live long enough
  --> tests/compile-fail/invalid_drop_counted.rs:13:37
   |
13 |             arena.alloc(DropCounted(&cell));
   |                                     ^^^^^ borrowed value does not live long enough
14 |         }
15 |     }
   |     - `cell` dropped here while still borrowed
...
21 | }
   | - borrow might be used here, when `arena` is dropped and runs the `Drop` code for type `DynamicArena`
   |
   = note: values in a scope are dropped in the opposite order they are defined
